File: algorithms/algorithms-cplusplus/leetcode/Question_0506_Relative_Ranks.cc
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  vector<string> findRelativeRanks(vector<int>& score) {
    vector<string> res;
    if (score.size() == 0) {
      return res;
    }
    vector<int> nums(score.begin(), score.end());
    sort(nums.rbegin(), nums.rend());
    unordered_map<int, int> m;
    for (int i = 0; i < nums.size(); i++) {
      m[nums[i]] = i;
    }
    for (int i = 0; i < score.size(); i++) {
      int index = m[score[i]];
      if (index == 0) {
        res.push_back("Gold Medal");
      } else if (index == 1) {
        res.push_back("Silver Medal");
      } else if (index == 2) {
        res.push_back("Bronze Medal");
      } else {
        res.push_back(to_string(index + 1));
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0506_Relative_Ranks.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.*;

public class Question_0506_Relative_Ranks {
    public String[] findRelativeRanks(int[] score) {
        if (score == null || score.length == 0) {
            return new String[0];
        }
        int n = score.length;
        int[] nums = Arrays.copyOf(score, score.length);
        Arrays.sort(nums);
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(nums[i], n - i - 1);
        }
        List<String> list = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            int index = map.get(score[i]);
            if (index == 0) {
                list.add("Gold Medal");
            } else if (index == 1) {
                list.add("Silver Medal");
            } else if (index == 2) {
                list.add("Bronze Medal");
            } else {
                list.add(index + 1 + "");
            }
        }
        return list.toArray(new String[0]);
    }
}
